stock_putaway_storage_type_strategy/models/stock_location.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2019 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import api, models, fields
+from odoo.osv.expression import AND, OR
+
+
+class StockLocation(models.Model):
+
+    _inherit = 'stock.location'
+
+    pack_storage_strategy = fields.Selection(
+        selection=[('none', 'None'),
+                   ('ordered_locations', 'Ordered Children Locations')],
+        required=True,
+        default='none',
+        string='Packs storage strategy',
+        help='TODO',
+    )
+    display_pack_storage_strategy = fields.Boolean(
+        compute='_compute_display_pack_storage_strategy'
+    )
+
+    @api.depends(
+        'child_ids', 'location_id', 'location_id.pack_storage_strategy',
+        'location_id.display_pack_storage_strategy'
+    )
+    def _compute_display_pack_storage_strategy(self):
+        for location in self:
+            # Do not display on locations without children
+            if not location.child_ids:
+                location.display_pack_storage_strategy = False
+                continue
+            current_location = location.location_id
+            display = True
+            while current_location:
+                # If a parent has pack_storage_strategy set, we don't display
+                # it on its children
+                if current_location.pack_storage_strategy != 'none':
+                    display = False
+                    break
+                current_location = current_location.location_id
+            location.display_pack_storage_strategy = display
+
+    def _get_putaway_rule_location(self, product=None, category=None):
+        """Get the destination location from putaway rules"""
+        quant = self.env.context.get('putaway_quant')
+        package_storage_type = False
+        if quant:
+            package_storage_type = quant.package_id.packaging_id.stock_package_storage_type_id
+        if not package_storage_type:
+            return super()._get_putaway_rule_location(
+                product=product, category=category
+            )
+        putaway_rules = self.putaway_rule_ids.filter_rules(
+            product=product, category=category
+        )
+        if category and not product:
+            product = quant.product_id
+        for rule in putaway_rules:
+            dest_location = rule._get_destination_location()
+            if (
+                dest_location.pack_storage_strategy == 'none' and
+                dest_location._package_storage_type_allowed(
+                    package_storage_type, quant, product=product
+                )
+            ):
+                return dest_location
+            storage_locations = dest_location.get_storage_locations(
+                product=product
+            )
+            allowed_location = storage_locations.select_first_allowed_location(
```

Comment:
No need to check locations one by one. "Area" contains the allowed storage type.